lib/AL_USDMaya/AL/usdmaya/nodes/LayerManager.h

```diff
@@ -0,0 +1,217 @@
+//
+// Copyright 2017 Animal Logic
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.//
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#pragma once
+#include "AL/usdmaya/Common.h"
+#include "AL/maya/NodeHelper.h"
+#include "pxr/pxr.h"
+#include "pxr/usd/usd/stage.h"
+
+#include "maya/MPxLocatorNode.h"
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/random_access_index.hpp>
+#include <boost/multi_index/hashed_index.hpp>
+
+#include <map>
+#include <set>
+#include <boost/thread.hpp>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace AL {
+namespace usdmaya {
+namespace nodes {
+
+
+//----------------------------------------------------------------------------------------------------------------------
+/// \brief  The layer node stores a reference to an SdfLayer.
+/// \ingroup nodes
+//----------------------------------------------------------------------------------------------------------------------
+class LayerManager
+  : public MPxNode,
+    public maya::NodeHelper
+{
+public:
+
+  /// \brief  ctor
+  inline LayerManager()
+    : MPxNode(), NodeHelper() {}
+
+  /// \brief  Find the already-existing non-referenced LayerManager node in the scene, or return a null MObject
+  /// \return the found LayerManager node, or a null MObject
+  static MObject findNode();
+
+  /// \brief  Either find the already-existing non-referenced LayerManager node in the scene, or make one
+  /// \return the found-or-created LayerManager node
+  static MObject findOrCreateNode();
+
+  /// \brief  Find the already-existing non-referenced LayerManager node in the scene, or return a nullptr
+  /// \return the found LayerManager, or a nullptr
+  static LayerManager* findManager();
```

Comment:
Can we have one method, findOrCreateManager()? and remove findManager? If we a buying into a singleton, I'd expect it to be created once the first time I access the LayerManager, or straight up create a LayerManager Node  when the AL_USDMaya plugin is loaded and we have a scene. Otherwise I think havint the two methods can complicate the code, since it seems like some code should call findOrCreateManager and some code should call findManager, but who really should create it?

Reply:
Not sure I completely follow what you mean by this:

"some code should call findOrCreateManager and some code should call findManager, but who really should create it?"

As for why I want two methods - basically, findManager() doubles as a "doesManagerExist()" query.  Yes, we could take the approach of always force-creating a manager whenever we do any sort of querying (or even on plugin load), but I'm strongly against this, because this involves creating a node in the scene. That then means that your scene now requires this plugin, even if it really doesn't - ie, if you have AL_USDMaya, open a new scene, then save out a blank scene, it now has a requirement to load AL_USDMaya.

We've had a few plugins from third-party vendors that operated like this, and they're highly annoying - the plugin ends up acting like a virus that infects scenes: when anyone opens such a scene, then opens and saves out another scene - even if that other scene doesn't need the plugin at all - they spread the "infection" to the other scene.

In our two studios, we may not care much, because I'm guessing we'll both nearly always want this plugin loaded, anyway, but it's a sort of good plugin-hygiene to only create nodes if the scene is question actually needs them.

Reply:
I guess I was seeing the LayerManager as the style of singleton where if you "access it, you create it" because otherwise you  will have 2 methods as you have, a FindAndCreate and a Find, which I think can cause confusion about when the creation of the LayerManager should happen. i.e. when should I call FindAndCreate?

But I see why you have done it this way, because you only want to create it when there is actually a ProxyShape in the scene which seems fair enough!